text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,460 @@
+
+- `FailedRetryable` - processing of telemetry data failed. The client SHOULD record the error and may retry exporting the same data. This can happen when the server is temporarily unable to process the data.
+
+#### Throttling
+
+OTLP allows backpressure signalling.
+
+If the server is unable to keep up with the pace of data it receives from the client then it SHOULD signal that fact to the client. The client MUST then throttle itself to avoid overwhelming the server.
+
+`Export` response includes a `throttle_period_millisec` field, which indicates the minimum duration in milliseconds that the client MUST wait before sending the next `Export` request. The waiting period starts from the moment the client receives such `Export` response. The value of 0 for `throttle_period_millisec` indicates that the client may send the next `Export` request immediately and is an indication that no throttling is requested by the server.
+
+This field is used by the server to signal backpressure to the client. The value of `throttle_period_millisec` is determined by the server and is implementation dependant. The server SHOULD choose a `throttle_period_millisec` value that is big enough to give the server time to recover, yet is not too big to cause the client to drop data while it is throttled.
+
+### gRPC Transport
+
+#### Service Definition
+
+`Hello` request and response is done via a unary gRPC call.
+
+`Export` requests and responses are delivered using bidirectional gRPC streams.
+
+This is OTLP over gRPC Service definition:
+
+```
+service StreamExporter {
+  // Hello handshake.
+  rpc Hello(HelloRequest) returns (HelloResponse) {}
+
+  // Sends a batch of telemetry data.
+  rpc Export(stream ExportRequest) returns (stream ExportResponse) {}
+}
+```
+
+Appendix A contains Protocol Buffer definitions for `HelloRequest`, `HelloResponse`, `ExportRequest` and `ExportResponse`.
+
+#### Load Balancing
+
+OTLP utilizes gRPC stream in a way that allows Level 7 load balancers that may be in the network path between the client and the server to re-balance the traffic periodically. This is necessary so that the traffic is not pinned by the load balancer to one server for the entire duration of telemetry data sending, thus potentially leading to imbalanced load of servers located behind the load balancer.
+
+OTLP achieves this by periodically closing and reopening the gRPC stream. This behavior is based on the fact that Level 7 load balancers that are gRPC-aware make re-balancing decisions when a new stream is opened.
+
+After sending an `Export` request the client MUST check how much time passed since the stream was last opened and if it is longer than a configurable period of time the stream MUST be closed and opened again. After opening the stream the client SHOULD perform a `Hello` message exchange.
```

Comment:
If this functionality is not needed it can be disabled on the client-side (by setting timeout period to infinity).

We can recommend to disable rebalancing in client libraries by default, because they are most likely sending to local daemon.

Agents and Collectors will have this enabled by default.